New slacksoc time!

brenns10 opened the pull request:
New Dockerfile. Tested with:

```bash
docker build .
docker run -v $(pwd)/state:/var/slacksoc --env-file=secrets.env IMAGE_ID
```

The `state` directory is where the state file will be saved. `secrets.env` has Github, Slack, and CWRU Love API keys with the following names:

```
SLACK_TOKEN
GITHUB_CLIENT_ID
GITHUB_CLIENT_SECRET
GITHUB_ACCESS_TOKEN
LOVE_API_KEY
```

The existing deployment of slacksoc should have suitable values for all, except LOVE_API_KEY, which I can provide

ajm188 commented on bot.go:
```diff
@@ -1,102 +0,0 @@
-	"os"
-
-	"github.com/ajm188/slack"
-	"github.com/ajm188/slack/plugins/github"
```
womp womp :(

brenns10 replied:
my apologies :crying_cat_face:

ajm188 replied:
Hey, it had to `go` (see what I did there? 😛 😂 )

ajm188 commented on slacksoc.yaml:
```diff
@@ -0,0 +1,18 @@
+statefile: /var/slacksoc/state.gob
+plugins:
```
okay this is awesome

brenns10 commented:
It would appear I forgot to update the travis config. I don't actually have a test suite, so the current travis config will always be angry with me.

matthewbentley commented:
Re: secrets (IDK if I can comment on a whole review?):

I'll look into using Docker Second (well, third I think) version of HacSoc hosting, but for now everything uses env vars.  Realistically anyone who has the ability to `inspect` a container can also read the env file right now.

matthewbentley commented on .gitignore:
```diff
@@ -0,0 +1 @@
+secrets.env
```
You should probably add a `.dockerignore` with the same contents, to keep your secret file from being copied in.

brenns10 commented:
Okay so hopefully the travis thing should be fixed (we'll see). I also made some updates to my slacksoc lib that puts some spammy debugging commands behind a list of "trusted" users (i.e. us). And added a PM me command.

brenns10 commented:
lol so not only is travis still failing, but I can't see the job log to verify anything

brenns10 commented:
Testing the slacksoc library is definitely something that I want to consider, but I don't think travis-ci builds necessarily belong here any more. This is really just a repository containing deployment stuff now. Any objections to removing travis-ci here, and placing an issue on brenns10/slacksoc about testing?

ajm188 commented:
I'm in favor of disabling travis. Now that @brenns10 has removed all the beautiful code I wrote, there's sort of nothing to test.

brenns10 commented:
Okay, so I made a major update here. I've put a stub `main.go` in (the exact same one from `github.com/brenns10/slacksoc/slacksoc`). This makes this repository `go get` -able once more.

Next, I used vendetta to create a vendor directory with all of the dependencies as submodules. As things change, I need to only do `vendetta -p` to remove old dependencies, and `vendetta -u` to update the submodules to their latest versions. Then commit the new submodules.

When you go get this, thanks to Go >=1.6 having vendoring by default, Go will use the submodule versions we have here, rather than getting the latest versions. Thus, the only way any of the code will change is if we bump the submodules with a commit.

brenns10 commented:
The Dockerfile in this PR won't work out of the box, because it is trying to `go get github.com/hacsoc/slacksoc`, which will get the master branch rather than this branch.

I could copy the whole source (and vendor) tree over in the dockerfile. Not sure whether that's a good idea though. Thoughts?

brenns10 commented:
I addressed the issue on my previous comment. You should be able to `git checkout`, `git submodule init --update`, and then build a docker image that runs just fine. I think we're ready once more.

ajm188 commented on README.md:
```diff
@@ -1,23 +1,25 @@
+
+## Updating Dependencies
+
+The convenient way to update all submodules is to use [vendetta][]. Install it
```
TIL you can do an implicit link ref if the link text matches. Who knew?

ajm188 commented:
Oh, and about the `COPY` vs `go get`:

You might get better docker cache hits with `COPY` (since the vendored files will always be identical unless you fetched new versions, in which case you're knowingly busting the cache). I'm not sure if docker will do some sort of checking on a digest of `go get` and realize a vendored package hasn't changed?

Does that make sense? I was really struggling with how to convey my concern with the `go get` case.

I suppose the best way would be to do `docker build` twice using the `go get` option and see if that layer comes from the cache the second time.

brenns10 commented:
Okay, updated the readme.
> You might get better docker cache hits with `COPY`
Yeah, I noticed that when I switched to `COPY`. When I use `COPY` the whole build is cached, so the second time I `docker build` I get the exact same hash. When I use `go get` it resumes from the `go get` step. I wonder if Docker can detect when socket system calls are made, and so it knows the build will depend on external resources.

brenns10 commented:
:confetti_ball:
